### ISSUE TITLE
Add more math functions

### ISSUE THREAD

**Issue by tylermenezes** [sass/sass#851]
*(edited by @nex3)*

* [x] [Proposal](https://github.com/sass/sass/pull/2779)
* [x] [Specs](https://github.com/sass/sass-spec/pull/1508)
* [x] [Dart Sass](https://github.com/sass/dart-sass/issues/897)
* [ ] [LibSass](https://github.com/sass/libsass/issues/3044)
* [x] [Documentation](https://github.com/sass/sass-site/issues/418)

---

Sass, by default, has a very limited set of math functions. Some functions which are missing, but would be of use:
- sqrt
- pow
- mod
- log

This functionality can be added with third-party gems, but the fact that they're missing from the core just makes everything more difficult out-of-the-box.

There is very little benefit to keeping mathematical functions out of the core, because they have a standard, expected definition, so they're not really cluttering the namespace.

**Comment by KittyGiraudel** [comment]
References:
- https://github.com/Team-Sass/Sassy-math
- http://compass-style.org/reference/compass/helpers/math/

**Comment by chriseppstein** [comment]
:thumbsup:

Note: modulo operations are available in Sass via the `%` operator.

**Comment by chriseppstein** [comment]
@nex3 do you object?

**Comment by ghost** [comment]
If we're going to add a log function, there should probably be log10 and log2 functions for convenience. Ordinary log should be base-e.

**Comment by nex3** [comment]
I want to wait on this until we have a better module story, since once we do it would be nice to put nice functions like this in a built-in module that can be imported with a prefix.

**Comment by booleanbetrayal** [comment]
:+1: 

would love to to see expanded math functionality that mimics the compass implementation

**Comment by torkiljohnsen** [comment]
:+1: Right now it's a pain to do stuff like `pow($x, $y)` if `$y` is a floating point number.

**Comment by KittyGiraudel** [comment]
Any chance of seeing this open again?

**Comment by Nevraeka** [comment]
@HugoGiraudel :+1: We really need the compass functionality built in at the sass level. LibSass won't add this functionality until its standard in sass. We need better consistency across the projects in that respect. @nex3 what are the updates on the module system?

**Comment by nex3** [comment]
All feature work is currently on hold until Libsass reaches compatibility.

**Comment by ixisio** [comment]
+1

**Comment by c9s** [comment]
Hi, this is planed in C6 http://github.com/c9s/c6

**Comment by ixisio** [comment]
@c9s It is not that easy to add another language to your running server and development infrastructure. So, if I plusone here, I will see the feature landing in libsass, not smth other frame.. whatever! Thanks sir!

**Comment by c9s** [comment]
@ixisio go libraries are actually dynamic linkable for any languages. :)

**Comment by nex3** [comment]
@c9s It is not remotely appropriate to come into a competitor's issue tracker solely to advertise your own technology. What you're doing is essentially spam. I've deleted your comments, and if you post more in this vein I won't hesitate to ban you from the repository entirely.

**Comment by dawidk92** [comment]
It would be nice to add trigonometric functions

+1

**Comment by dazld** [comment]
Really missing trig / math functions when using libsass, and the bottleneck seems to be here..  understand there is a lot going on, but if possible please do prioritise these valuable additions. 

Generic math seems to me to be in a different category to other more specific functions that would belong in a custom extension.

**Comment by mootari** [comment]
@nex3 According to http://sass-compatibility.github.io/ libsass 3.3 has reached full feature parity. What compatibility issues that block this issue are still left?

**Comment by nex3** [comment]
It's not compatibility that's blocking this, it's getting a module system in place.

**Comment by chriseppstein** [comment]
Note that for node-sass users, using eyeglass + the eyeglass-math module will get you all of this functionality in the interim.

**Comment by cherscarlett** [comment]
I'm definitely going to +1 this here. I'm currently trying to get a number of colleagues to use SASS over LESS and these are core features in LESS that they'd need to utilize eyeglass + eyeglass-math to use.

**Comment by dom1n1k** [comment]
+1

**Comment by pdf13** [comment]
+1

**Comment by lunelson** [comment]
Why is the module system required for this actually? Because of existing codebases which may have already assigned functions to those names? In those cases the native functions would be just be overridden, but if those users remove their custom functions, the native ones would come through wouldn't they? 

I would agree that much of JS Math is pretty edge-casey for CSS but one that is really important is `Math.pow()` for gamma compensation: there are many functions in the wild for trying to evaluate WCAG [relative-luminance](https://www.w3.org/TR/WCAG20/#relativeluminancedef) and subsequent contrast-ratios, (notably [Bootstrap's `yiq()` Sass function](https://github.com/twbs/bootstrap/blob/ff29c1224c20b8fcf2d1e7c28426470f1dc3e40d/scss/_functions.scss#L52-L64)) _which are doing it incorrectly_. Relative luminance calculation is impossible without `pow()`.

**Comment by nex3** [comment]
The module system is required because we don't want to add top-level functions in one release only to deprecate them a few releases later; and because terse top-level functions run a high risk of conflicting with future CSS additions in ways that require lots of complexity to deal with (as we've already seen with `min()` and `max()`). In fact, if we'd added `pow()`, [we would have run into exactly that issue](https://twitter.com/bdc/status/1100921258839953408).

If you're impatient to see the module system land, there's lots you could do that would help accelerate it, even without directly writing code. For example, one of my big tasks this quarter is fleshing out some tests in sass-spec; if someone else would pitch in to help with that, I could focus more time on implementing modules.

**Comment by lunelson** [comment]
@nex3 Wow! I never thought we'd see native math functions in CSS. I understand the point then—I've even found I need to suppress' Sass rgb() function in certain custom property use-cases—is there a document you can point me to, that explains where and how to help with sass-spec?

**Comment by nex3** [comment]
@lunelson The documentation in the repo could use some love. Let me update it and get back to you.

**Comment by nex3** [comment]
[This pull request](https://github.com/sass/sass-spec/pull/1353) should add enough documentation for you to get started. Let me know if there's anything it doesn't explain adequately.

For function specs in particular, [the `variable-exists()` spec](https://github.com/sass/sass-spec/blob/master/spec/core_functions/variable_exists.hrx) is probably a good template to give you an idea of what they should look like. Every function Sass provides that doesn't have a spec yet needs one. You can look at [the Ruby Sass function tests](https://github.com/sass/ruby-sass/blob/stable/test/sass/functions_test.rb) to get an idea of all the cases that need to be covered.

**Comment by lunelson** [comment]
Thanks @nex3 I'm going to try to wade in to it. Is there a priority list of some kind, as to which tests are most important for your work to progress? And/Or is there an issue/PR protocol to follow, wrt choosing what stuff to work on?

**Comment by nex3** [comment]
The goal is just to pay off technical debt, so all functions are pretty
much equal priority.

**Comment by nex3** [comment]
The module system has landed, so I'm officially marking this "planned". I'd like to expose Sassified versions of all the mathematical functions in the [Values and Units 4 Draft](https://drafts.csswg.org/css-values-4/#math), modulo, and logarithms (probably as a single function with an optional `$base` argument that defaults to e). I'd also like to expose `$e` and `$pi` variables.

There are some interesting questions about what these functions should do with units; fortunately, the CSS working group has done a lot of the hard work of figuring that out for us already in Values and Units 4! The only remaining questions are modulo (which should require its arguments to have compatible units and return the same unit) and logarithms (which I think should require unitless arguments).

**Comment by nex3** [comment]
@Awjin's [math function proposal](https://github.com/sass/sass/blob/master/proposal/more-math-functions.md) has landed and is now officially open for comment. I'll leave it open for a couple weeks before marking it as accepted.

**Comment by nex3** [comment]
Closing this out because [LibSass is now deprecated](https://sass-lang.com/blog/libsass-is-deprecated) and we aren't expecting to add any additional features to it.
